layerindex/update_layer.py

```diff
@@ -55,8 +58,49 @@ def split_recipe_fn(path):
         pv = "1.0"
     return (pn, pv)
 
-def update_recipe_file(tinfoil, data, path, recipe, layerdir_start, repodir):
+def collect_patch(recipe, patchfn, index, layerdir_start, stop_on_error):
+    from django.db import DatabaseError
+    from layerindex.models import Patch
+
+    patchrec = Patch()
+    patchrec.recipe = recipe
+    patchrec.path = os.path.relpath(patchfn, layerdir_start)
+    patchrec.src_path = os.path.relpath(patchrec.path, recipe.filepath)
+    patchrec.apply_order = index
+    try:
+        patchrec.read_status_from_file(patchfn, logger)
+        patchrec.save()
+    except DatabaseError:
+        raise
+    except Exception as e:
+        if stop_on_error:
+            raise
+        else:
+            logger.error("Unable to read patch %s: %s", patchfn, str(e))
+            patchrec.save()
+
+def collect_patches(recipe, envdata, layerdir_start, stop_on_error):
+    from layerindex.models import Patch
+
+    try:
+        import oe.recipeutils
+    except ImportError:
+        logger.warn('Failed to find lib/oe/recipeutils.py in layers - patches will not be imported')
+        return
+
+    Patch.objects.filter(recipe=recipe).delete()
+    patches = oe.recipeutils.get_recipe_patches(envdata)
+    for i, patch in enumerate(patches):
+        if not patch.startswith(layerdir_start):
+            # Likely a remote patch, skip it
+            continue
+        collect_patch(recipe, patch, i, layerdir_start, stop_on_error)
+
+def update_recipe_file(tinfoil, data, path, recipe, layerdir_start, repodir, stop_on_error, skip_patches=False):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function contains 72 lines of code, not including blank lines or lines with only comments, Python punctuation characters, identifiers, or literals. By comparison, 99% of the functions in the CodeGuru reference dataset contain fewer lines of code. Large functions might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this function or break it into multiple functions.